local Random = Random.new()

print(({
    --/ https://patorjk.com/software/taag
    [[

                 .-.                                
                /    \                              
  ___ .-. .-.   | .`. ;    .--.     .--.     .--.   
 (   )   '   \  | |(___)  /    \   /    \   /    \  
 |  .-.  .-. ;  | |_     |  .-. ; |  .-. ; |  .-. ; 
 | |  | |  | | (   __)   |  | | | |  | | | |  | | | 
 | |  | |  | |  | |      |  |/  | |  |/  | |  |/  | 
 | |  | |  | |  | |      |  ' _.' |  ' _.' |  ' _.' 
 | |  | |  | |  | |      |  .'.-. |  .'.-. |  .'.-. 
 | |  | |  | |  | |      '  `-' / '  `-' / '  `-' / 
(___)(___)(___)(___)      `.__.'   `.__.'   `.__.'  
]],
    [[

     ______  _______         _____       ______        ______        ______   
    |      \/       \   ____|\    \  ___|\     \   ___|\     \   ___|\     \  
   /          /\     \ |    | \    \|     \     \ |     \     \ |     \     \ 
  /     /\   / /\     ||    |______/|     ,_____/||     ,_____/||     ,_____/|
 /     /\ \_/ / /    /||    |----'\ |     \--'\_|/|     \--'\_|/|     \--'\_|/
|     |  \|_|/ /    / ||    |_____/ |     /___/|  |     /___/|  |     /___/|  
|     |       |    |  ||    |       |     \____|\ |     \____|\ |     \____|\ 
|\____\       |____|  /|____|       |____ '     /||____ '     /||____ '     /|
| |    |      |    | / |    |       |    /_____/ ||    /_____/ ||    /_____/ |
 \|____|      |____|/  |____|       |____|     | /|____|     | /|____|     | /
    \(          )/       )/           \( |_____|/   \( |_____|/   \( |_____|/ 
     '          '        '             '    )/       '    )/       '    )/    
                                            '             '             '     
]],
    [[

 ___      ___   _______    _______    _______    _______  
|"  \    /"  | /"     "|  /"     "|  /"     "|  /"     "| 
 \   \  //   |(: ______) (: ______) (: ______) (: ______) 
 /\\  \/.    | \/    |    \/    |    \/    |    \/    |   
|: \.        | // ___)    // ___)_   // ___)_   // ___)_  
|.  \    /:  |(:  (      (:      "| (:      "| (:      "| 
|___|\__/|___| \__/       \_______)  \_______)  \_______) 
]],
    [[

<-. (`-')              (`-')  _  (`-')  _  (`-')  _ 
   \(OO )_    <-.      ( OO).-/  ( OO).-/  ( OO).-/ 
,--./  ,-.)(`-')-----.(,------. (,------. (,------. 
|   `.'   |(OO|(_\---' |  .---'  |  .---'  |  .---' 
|  |'.'|  | / |  '--. (|  '--.  (|  '--.  (|  '--.  
|  |   |  | \_)  .--'  |  .--'   |  .--'   |  .--'  
|  |   |  |  `|  |_)   |  `---.  |  `---.  |  `---. 
`--'   `--'   `--'     `------'  `------'  `------' 
]],
    [[

 __  __      ___                             
|  \/  |    | __|    ___      ___      ___   
| |\/| |    | _|    / -_)    / -_)    / -_)  
|_|__|_|   _|_|_    \___|    \___|    \___|  
_|"""""| _| """ | _|"""""| _|"""""| _|"""""| 
"`-0-0-' "`-0-0-' "`-0-0-' "`-0-0-' "`-0-0-' 
]],
    [[

â–ˆâ–ˆâ–ˆâ•—   â–ˆâ–ˆâ–ˆâ•—â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—
â–ˆâ–ˆâ–ˆâ–ˆâ•— â–ˆâ–ˆâ–ˆâ–ˆâ•‘â–ˆâ–ˆâ•”â•â•â•â•â•â–ˆâ–ˆâ•”â•â•â•â•â•â–ˆâ–ˆâ•”â•â•â•â•â•â–ˆâ–ˆâ•”â•â•â•â•â•
â–ˆâ–ˆâ•”â–ˆâ–ˆâ–ˆâ–ˆâ•”â–ˆâ–ˆâ•‘â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—  â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—  â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—  â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—  
â–ˆâ–ˆâ•‘â•šâ–ˆâ–ˆâ•”â•â–ˆâ–ˆâ•‘â–ˆâ–ˆâ•”â•â•â•  â–ˆâ–ˆâ•”â•â•â•  â–ˆâ–ˆâ•”â•â•â•  â–ˆâ–ˆâ•”â•â•â•  
â–ˆâ–ˆâ•‘ â•šâ•â• â–ˆâ–ˆâ•‘â–ˆâ–ˆâ•‘     â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ•—
â•šâ•â•     â•šâ•â•â•šâ•â•     â•šâ•â•â•â•â•â•â•â•šâ•â•â•â•â•â•â•â•šâ•â•â•â•â•â•â•
]],
    [[

â–ˆâ–ˆâ–ˆâ–„ â–„â–„â–ˆâ–ˆâ–ˆâ–“ â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–’â–“â–ˆâ–ˆâ–ˆâ–ˆâ–ˆ â–“â–ˆâ–ˆâ–ˆâ–ˆâ–ˆ â–“â–ˆâ–ˆâ–ˆâ–ˆâ–ˆ 
â–“â–ˆâ–ˆâ–’â–€â–ˆâ–€ â–ˆâ–ˆâ–’â–“â–ˆâ–ˆ   â–’ â–“â–ˆ   â–€ â–“â–ˆ   â–€ â–“â–ˆ   â–€ 
â–“â–ˆâ–ˆ    â–“â–ˆâ–ˆâ–‘â–’â–ˆâ–ˆâ–ˆâ–ˆ â–‘â–’â–ˆâ–ˆâ–ˆ   â–’â–ˆâ–ˆâ–ˆ   â–’â–ˆâ–ˆâ–ˆ    
â–’â–ˆâ–ˆ    â–’â–ˆâ–ˆ â–‘â–“â–ˆâ–’  â–‘â–’â–“â–ˆ  â–„ â–’â–“â–ˆ  â–„ â–’â–“â–ˆ  â–„  
â–’â–ˆâ–ˆâ–’   â–‘â–ˆâ–ˆâ–’â–‘â–’â–ˆâ–‘   â–‘â–’â–ˆâ–ˆâ–ˆâ–ˆâ–’â–‘â–’â–ˆâ–ˆâ–ˆâ–ˆâ–’â–‘â–’â–ˆâ–ˆâ–ˆâ–ˆâ–’ 
â–‘ â–’â–‘   â–‘  â–‘ â–’ â–‘    â–‘â–‘ â–’â–‘ â–‘â–‘â–‘ â–’â–‘ â–‘â–‘â–‘ â–’â–‘ â–‘
â–‘  â–‘      â–‘ â–‘       â–‘ â–‘  â–‘ â–‘ â–‘  â–‘ â–‘ â–‘  â–‘
â–‘      â–‘    â–‘ â–‘       â–‘      â–‘      â–‘   
    â–‘              â–‘  â–‘   â–‘  â–‘   â–‘  â–‘   
]],
    [[

     e    e      888~~                                
    d8b  d8b     888___  e88~~8e   e88~~8e   e88~~8e  
   d888bdY88b    888    d888  88b d888  88b d888  88b 
  / Y88Y Y888b   888    8888__888 8888__888 8888__888 
 /   YY   Y888b  888    Y888    , Y888    , Y888    , 
/          Y888b 888     "88___/   "88___/   "88___/  
]],
    [=[

.        :   .-:::::'                              
;;,.    ;;;  ;;;''''                               
[[[[, ,[[[[, [[[,,== ,cc[[[cc. ,cc[[[cc. ,cc[[[cc. 
$$$$$$$$"$$$ `$$$"`` $$$___--' $$$___--' $$$___--' 
888 Y88" 888o 888    88b    ,o,88b    ,o,88b    ,o,
MMM  M'  "MMM "MM,    "YUMMMMP" "YUMMMMP" "YUMMMMP"
]=],
    [[
    
--------------
4D 46 65 65 65
--------------
]],
    [[

     _         _         _         _         _    
   _( )__    _( )__    _( )__    _( )__    _( )__ 
 _|     _| _|     _| _|     _| _|     _| _|     _|
(_ M _ (_ (_ F _ (_ (_ e _ (_ (_ e _ (_ (_ e _ (_ 
  |_( )__|  |_( )__|  |_( )__|  |_( )__|  |_( )__|
]],
    [[

_______________________________________________________
/~~\__/~~\_/~~~~~~~~\_/~~~~~~~~\_/~~~~~~~~\_/~~~~~~~~\_
/~~~\/~~~\_/~~\_______/~~\_______/~~\_______/~~\_______
/~~~~~~~~\_/~~~~~~\___/~~~~~~\___/~~~~~~\___/~~~~~~\___
/~~\__/~~\_/~~\_______/~~\_______/~~\_______/~~\_______
/~~\__/~~\_/~~\_______/~~~~~~~~\_/~~~~~~~~\_/~~~~~~~~\_
_______________________________________________________
]],
    [[

\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
/// \\/// \///////// \///////// \///////// \///////// \
//// //// \/// \\\\\\\/// \\\\\\\/// \\\\\\\/// \\\\\\\
/// / /// \/////// \\\/////// \\\/////// \\\/////// \\\
/// \\/// \/// \\\\\\\/// \\\\\\\/// \\\\\\\/// \\\\\\\
/// \\/// \/// \\\\\\\///////// \///////// \///////// \
\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
]],
    [[

 _______  _______                      
(_______)(_______)                     
 _  _  _  _____    _____  _____  _____ 
| ||_|| ||  ___)  | ___ || ___ || ___ |
| |   | || |      | ____|| ____|| ____|
|_|   |_||_|      |_____)|_____)|_____)
]],
    [[

______________________________________
7        77     77     77     77     7
|  _  _  ||  ___!|  ___!|  ___!|  ___!
|  7  7  ||  __| |  __|_|  __|_|  __|_
|  |  |  ||  7   |     7|     7|     7
!__!__!__!!__!   !_____!!_____!!_____!
]],
    [[

     _/\/\______/\/\__/\/\/\/\/\/\_____________________________________
    _/\/\/\__/\/\/\__/\/\____________/\/\/\______/\/\/\______/\/\/\___ 
   _/\/\/\/\/\/\/\__/\/\/\/\/\____/\/\/\/\/\__/\/\/\/\/\__/\/\/\/\/\_  
  _/\/\__/\__/\/\__/\/\__________/\/\________/\/\________/\/\_______   
 _/\/\______/\/\__/\/\____________/\/\/\/\____/\/\/\/\____/\/\/\/\_    
__________________________________________________________________     
]],
    [[

_/\/\______/\/\__/\/\/\/\/\/\_____________________________________
_/\/\/\__/\/\/\__/\/\____________/\/\/\______/\/\/\______/\/\/\___
_/\/\/\/\/\/\/\__/\/\/\/\/\____/\/\/\/\/\__/\/\/\/\/\__/\/\/\/\/\_
_/\/\__/\__/\/\__/\/\__________/\/\________/\/\________/\/\_______
_/\/\______/\/\__/\/\____________/\/\/\/\____/\/\/\/\____/\/\/\/\_
__________________________________________________________________
]],
    [[

$$\      $$\ $$$$$$$$\                               
$$$\    $$$ |$$  _____|                              
$$$$\  $$$$ |$$ |       $$$$$$\   $$$$$$\   $$$$$$\  
$$\$$\$$ $$ |$$$$$\    $$  __$$\ $$  __$$\ $$  __$$\ 
$$ \$$$  $$ |$$  __|   $$$$$$$$ |$$$$$$$$ |$$$$$$$$ |
$$ |\$  /$$ |$$ |      $$   ____|$$   ____|$$   ____|
$$ | \_/ $$ |$$ |      \$$$$$$$\ \$$$$$$$\ \$$$$$$$\ 
\__|     \__|\__|       \_______| \_______| \_______|
]]
})[Random:NextInteger(1, 18)])

--// Festivals and Emojis //
local Year = tonumber(os.date("%Y")) :: number
local Easter, Mother, Father

local function Calculate(First: number, Second: number, Month: number): string?
    for Day = First, Second do
        if os.date("%w", os.time({year = Year, month = Month, day = Day})) == "0" then
            return `{Month} {Day}`
        end
    end
    return
end

do
    local A = math.floor(Year / 100)
    local B = math.floor((13 + 8 * A) / 25)
    local C = (15 - B + A - math.floor(A / 4)) % 30
    local D = (4 + A - math.floor(A / 4)) % 7
    local E = (19 * (Year % 19) + C) % 30
    local F = (2 * (Year % 4) + 4 * (Year % 7) + 6 * E + D) % 7
    local G = (22 + E + F)

    Easter = (
        (E == 29 and F == 6 and "04 19") or
        (E == 28 and F == 6 and "04 18") or
        (31 < G and (`04 {G - 31}`)) or
        (`03 {G}`)
    )
    Mother = Calculate( 8, 14, 5)
    Father = Calculate(15, 21, 6)
end

local EmojiList = {
    ["01 01"] = "ðŸŽ‡",
    ["02 14"] = ({"ðŸ’š", "ðŸ’›", "ðŸ§¡", "â¤ï¸", "ðŸ’—"})[Random:NextInteger(1, 5)],
    ["03 08"] = "ðŸ‘©",
    ["03 12"] = "ðŸŒ³",
    [Easter ] = "ðŸ¥š",
    ["04 01"] = utf8.char(Random:NextInteger(0x1F600, 0x1F64F)),
    [Mother ] = "ðŸ‘©",
    ["06 01"] = "ðŸ‘¶",
    [Father ] = "ðŸ‘¨",
    ["10 29"] = "ðŸŽƒ",
    ["10 30"] = "ðŸŽƒ",
    ["10 31"] = "ðŸŽƒ",
    ["11 01"] = "ðŸŽƒ",
    ["12 24"] = "ðŸŽ„",
    ["12 25"] = "ðŸŽ„"
} :: {[string?]: string}

if wax.shared.Language == "zh-cn" then
    EmojiList["01 26"] = "ðŸ¥£" --/ 2026
    EmojiList["02 16"] = "ðŸ§¨" --/ 2026
    EmojiList["03 03"] = "ðŸ®" --/ 2026
    EmojiList["05 31"] = "ðŸƒ" --/ 2025
    EmojiList["10 06"] = "ðŸ¥®" --/ 2025
    EmojiList["04 04"] = "ðŸª¦"
    EmojiList["04 05"] = "ðŸª¦"
    EmojiList["04 06"] = "ðŸª¦"
    EmojiList["10 06"] = "ðŸ¥®" --/ 2025

    for Date = 17, 22 do
        EmojiList[`02 {Date}`] = "ðŸ§§" --/ 2026
    end

    for Date = 1, 5 do
        EmojiList[`05 {Date}`] = "ðŸ­"
    end
end

return EmojiList[os.date("%m %d")] or ""
